### ISSUE TITLE
newWithSameType pimp function implementation is not 100% safe

### ISSUE THREAD

**Issue by danielpetisme** [eclipse/golo-lang#8]
Currently most of the Collections/functional function pimps are based on the `newWithSameType` function. The function name is quite explicit but during the runtime we can face some disturbing issues. Indeed, some "Types" are not accessible for instanciation.  Private inner classes for instance, such as `java.util.HashMap$EntrySet
`. It's a private implementation of the `Set` interface.
http://grepcode.com/file/repository.grepcode.com/java/root/jdk/openjdk/6-b14/java/util/HashMap.java

There is a usecase based on the now famous Mr. Bean....

```
module test

local function mrbean = -> DynamicObject():
  name("Mr Bean"):
  email("mrbean@gmail.com"):
  toString(|this| -> this: name() + " <" + this: email() + ">")

function main = |args| {

  let bean = mrbean()
  println(bean: toString())

  bean: email("mrbean@outlook.com")
  println(bean: toString())

  println("Displaying all the properties")
  println(bean: properties())
  println(bean: properties(): getClass())
  println(bean: properties(): filter(|it| -> it: getKey() == "name"))

}

```

and there you have the statck:

```
Mr Bean <mrbean@gmail.com>
Mr Bean <mrbean@outlook.com>
Displaying all the properties
[email=mrbean@outlook.com, name=Mr Bean, toString=MethodHandle(Object)Object]
class java.util.HashMap$EntrySet
Exception in thread "main" java.lang.reflect.InvocationTargetException
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:601)
    at fr.insalyon.citi.golo.cli.MainGoloGolo.main(MainGoloGolo.java:69)
Caused by: java.lang.InstantiationException: java.util.HashMap$EntrySet
    at java.lang.Class.newInstance0(Class.java:359)
    at java.lang.Class.newInstance(Class.java:327)
    at java.lang.invoke.MethodHandleImpl$BindCaller$T/24300697.invoke_V(MethodHandleImpl.java:1404)
    at java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:566)
    at fr.insalyon.citi.golo.runtime.MethodInvocationSupport.fallback(MethodInvocationSupport.java:176)
    at gololang.StandardPimps$java$util$Collection.newWithSameType(standard-pimps.golo)
    at java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:566)
    at fr.insalyon.citi.golo.runtime.MethodInvocationSupport.fallback(MethodInvocationSupport.java:176)
    at gololang.StandardPimps$java$util$Set.filter(standard-pimps.golo)
    at java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:566)
    at fr.insalyon.citi.golo.runtime.MethodInvocationSupport.fallback(MethodInvocationSupport.java:176)
    at test.main(test1.golo)
    ... 5 more
```

So basically I'm trying to invoke the `filter` pimp to get only the "name" property (the original use case was to filter closures w/ `isClosure`). The `filter` pimp try to create a fresh instance of `java.util.HashMap$EntrySet` but it can not and BAZINGA!!

I don't think it could be consider as a language limitation since the behavior is unpredictable. We're not responsible of the implementation underneath an interface.
Of course, we could imagine setting up some reflection voodoo to be able to instantiate a private class but the SecurityManager policies won't help us to do so....
Maybe the easiest solution could be to define some default types which will be use instead of the original one if we"re not able to access/instantiate it.
